src/main/java/org/truffleruby/core/regexp/RegexpNodes.java

```diff
@@ -140,30 +150,15 @@ public static DynamicObject matchCommon(RubyContext context, Node currentNode, D
         return matchData;
     }
 
-    // because the factory is not constant
     @TruffleBoundary
-    private static DynamicObject createSubstring(RopeNodes.MakeSubstringNode makeSubstringNode, DynamicObject source, int start, int length) {
-        assert RubyGuards.isRubyString(source);
-
-        final Rope sourceRope = StringOperations.rope(source);
-        final Rope substringRope = makeSubstringNode.executeMake(sourceRope, start, length);
-
-        final DynamicObject ret = Layouts.CLASS.getInstanceFactory(Layouts.BASIC_OBJECT.getLogicalClass(source)).newInstance(Layouts.STRING.build(false, false, substringRope));
-
-        return ret;
-    }
-
-    private static void setLocalVariable(Frame frame, String name, Object value) {
-        assert value != null;
-
-        while (frame != null) {
-            final FrameSlot slot = frame.getFrameDescriptor().findFrameSlot(name);
-            if (slot != null) {
-                frame.setObject(slot, value);
-                break;
-            }
-
-            frame = RubyArguments.getDeclarationFrame(frame);
+    private static int runMatch(RubyContext context, Node currentNode, Matcher matcher, int startPos, int range, boolean onlyMatchAtStart) {
+        // Keep status as RUN because MRI has an uninterruptible Regexp engine
+        if (onlyMatchAtStart) {
+            return context.getThreadManager().runUntilResultKeepStatus(currentNode,
+                    () -> matcher.matchInterruptible(startPos, range, Option.DEFAULT));
```

Comment:
Does `match` instead of `search` also implies the regexp has to match until the end of the String or not?

Reply:
Only if the regexp itself specifies that it must match to the end. For example
```
/my\z/.match_start "my, oh my",0
```
will return `nil`, but
```
/my/.match_start "my, oh my",0
```
will return `MatchData`

Reply:
OK :)